File: api/src/main/java/net/blueshell/api/config/HibernateDirtyTrackingConfig.java
package net.blueshell.api.config;

import net.blueshell.api.common.hibernate.DirtyTrackingInterceptor;
import org.hibernate.Interceptor;
import org.springframework.boot.autoconfigure.orm.jpa.HibernatePropertiesCustomizer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.util.Map;

/**
 * Registers the interceptor at the SessionFactory level.
 */
@Configuration
public class HibernateDirtyTrackingConfig {

    @Bean
    public Interceptor dirtyTrackingInterceptor() {
        return new DirtyTrackingInterceptor();
    }

    @Bean
    public HibernatePropertiesCustomizer dirtyTrackingCustomizer(Interceptor dirtyTrackingInterceptor) {
        return (Map<String, Object> props) ->
                props.put("hibernate.session_factory.interceptor", dirtyTrackingInterceptor);
    }
}

File: api/src/main/kotlin/net/blueshell/api/platform/config/HibernateDirtyTrackingConfig.kt
package net.blueshell.api.platform.config

import net.blueshell.api.shared.model.hibernate.DirtyTrackingInterceptor
import org.hibernate.Interceptor
import org.springframework.boot.autoconfigure.orm.jpa.HibernatePropertiesCustomizer
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

/**
 * Registers the interceptor at the SessionFactory level.
 */
@Configuration
class HibernateDirtyTrackingConfig {
    @Bean
    fun dirtyTrackingInterceptor(): Interceptor {
        return DirtyTrackingInterceptor()
    }

    @Bean
    fun dirtyTrackingCustomizer(dirtyTrackingInterceptor: Interceptor?): HibernatePropertiesCustomizer {
        return HibernatePropertiesCustomizer { props: MutableMap<String?, Any?>? ->
            props!!["hibernate.session_factory.interceptor"] = dirtyTrackingInterceptor
        }
    }
}
